languages/csharp/exercises/concept/time/.meta/Example.cs

```diff
@@ -0,0 +1,147 @@
+using System;
+using System.Globalization;
+using System.Runtime.InteropServices;
+
+public enum Location
+{
+    NewYork,
+    London,
+    Paris
+}
+
+public enum AlertLevel
+{
+    Early,
+    Standard,
+    Late
+}
+
+public static class Appointment
+{
+    public static DateTime ShowLocalTime(DateTime dtUtc)
+    {
+        return dtUtc.ToLocalTime();
+    }
+
+    public static DateTime Schedule(string appointmentDateDescription, Location location)
+    {
+        TimeZoneInfo tziLocation = TimeZoneInfo.FindSystemTimeZoneById(GetTimeZoneId(location));
+        DateTime dtl = DateTime.Parse(appointmentDateDescription);
+        DateTime dtu = TimeZoneInfo.ConvertTimeToUtc(dtl, tziLocation);
+        return dtu;
+    }
+
+    public static DateTime GetAlertTime(DateTime appointment, AlertLevel alertLevel)
+    {
+        TimeSpan noticePeriod = new TimeSpan();
+
+        switch (alertLevel)
+        {
+            case AlertLevel.Early:
+                noticePeriod = new TimeSpan(1, 0, 0, 0);
+                break;
+            case AlertLevel.Standard:
+                noticePeriod = new TimeSpan(1, 45, 0);
+                break;
+            case AlertLevel.Late:
+                noticePeriod = new TimeSpan(0, 30, 0);
+                break;
+        }
+
+        return appointment - noticePeriod;
+    }
+
+    public static bool HasDaylightSavingChanged(DateTime dt, Location location)
+    {
+        DateTime dtPrevious = dt.AddDays(-7);
+        string tzid = string.Empty;
+        TimeZoneInfo tzi = TimeZoneInfo.FindSystemTimeZoneById(GetTimeZoneId(location));
+        return tzi.IsDaylightSavingTime(dtPrevious) != tzi.IsDaylightSavingTime(dt);
+    }
+
+    public static DateTime NormalizeDateTime(string dtStr, Location location)
+    {
+        try
+        {
+            return DateTime.Parse(dtStr, LocationToCulture(location));
+        }
+        catch (Exception)
+        {
+            return DateTime.MinValue;
+        }
+    }
+
+    private static CultureInfo LocationToCulture(Location location)
+    {
+        string cultureId = string.Empty;
+        switch (location)
+        {
+            case Location.NewYork:
+                cultureId = "en-US";
+                break;
+            case Location.London:
+                cultureId = "en-GB";
+                break;
+            case Location.Paris:
+                cultureId = "fr-FR";
+                break;
+        }
+        return new CultureInfo(cultureId);
+    }
+
+    // ****** we are not exercising cross-platform skills as part of the exercise.
```

Comment:
We could consider having the time zone ID retrieval methods already stubbed out, because as you said, we are not teaching cross-platform skills. What do you think?

Reply:
> we are not teaching cross-platform skills

I agree that we don't want the student to start messing about with different platforms but I would suggest it is important that they are aware of this unhappiness relating to time zones and that they get hands-on with the ids which are really one the trickiest things in this area.  I want them to think about ids, that they come from outside of .NET and in the case of Windows are not a very nice format.

Assuming I have understood your position I would still push for them to be doing their own location-to-timezoneId mapping.

Reply:
> Assuming I have understood your position I would still push for them to be doing their own location-to-timezoneId mapping.

Sure. I'm wondering if we should mention in the `introduction.md` file that there is a `RuntimeInformation` class, just to get them looking in the right direction.

Reply:
> Sure. I'm wondering if we should mention in the introduction.md file that there is a RuntimeInformation class, just to get them looking in the right direction.

I'm a bit confused now.  My original idea (which the instructions and template code reflect) was that the student would not h be concerned with any cross-platform **technology** and we would provide appropriate time zone ids for their platform (in the instructions).  They would just be aware that there was a potential cross-platform issue.

I suppose there should be a note for students using the in-browser experience.  I'm assuming it is Linux.

I don't mind mentioning cross-platform technology or including it in the exercise.  I just want to put the students in a position where they have to take notice of these ids.  I think if we give them the ids in the code then they will forget about them.

I think we have a number of options including mentioning the `RuntimeInformation` class in the introduction but I would like to unconfused about your position and make sure you understand mine, first.

Reply:
> I'm a bit confused now. My original idea (which the instructions and template code reflect) was that the student would not h be concerned with any cross-platform technology and we would provide appropriate time zone ids for their platform (in the instructions). They would just be aware that there was a potential cross-platform issue.

Well, it isn't just a potential cross-platform issue, as for CLI users that will be using Windows, their tests will fail when run in the test runner. So we have to do something to get them to write the correct code.

Makes sense?

Reply:
I hadn't thought of that.  My understanding is now that even if it works on Windows on a local machine it will break when uploaded to the test runner unless some cross-platform capability is built in.

An alternative would be to put the cross-platform logic in the tests but I think that would be worse than in _Time.cs_.

Ok. I can't think of anything better than stubbing out the code and making a fuss about time zone ids in the _instructions.md_  as well as _after.md_.  Unsatisfactory, but that is life.  I will do that and mention the `RuntimeInformation` class in the introduction.

Reply:
> Ok. I can't think of anything better than stubbing out the code and making a fuss about time zone ids in the instructions.md as well as after.md. Unsatisfactory, but that is life. I will do that and mention the RuntimeInformation class in the introduction.

I'd be totally okay with just mentioning `RuntimeInformation` in the `introduction.md` and then having the students figuring it out themselves (with some relatively spoiler-y hints).

Reply:
So you're saying let the students do the cross-platform stuff.  No code mapping location to timezone id and no cross-platform code is given away.

That is fine with me.  They get to grapple with the time zone ids.

Do you want to add cross-platform as a concept?

Reply:
Yes, precisely. If we give them a very nice hint and mention `RuntimeInformation` in the `introduction.md`, I think most of the will be able to figure this out.

Reply:
Implemented the above.  See comment below.